File: mendeleev.py
import sys


# fmt: off
ELEMENTS = [
  "Ac", "Ag", "Al", "Am", "Ar", "As", "At", "Au", "B", "Ba", "Be", "Bh",
  "Bi", "Bk", "Br", "C", "Ca", "Cd", "Ce", "Cf", "Cl", "Cm", "Cn", "Co",
  "Cr", "Cs", "Cu", "Db", "Ds", "Dy", "Er", "Es", "Eu", "F", "Fe", "Fl",
  "Fm", "Fr", "Ga", "Gd", "Ge", "H", "He", "Hf", "Hg", "Ho", "Hs", "I",
  "In", "Ir", "K", "Kr", "La", "Li", "Lr", "Lu", "Lv", "Mc", "Md", "Mg",
  "Mn", "Mo", "Mt", "N", "Na", "Nb", "Nd", "Ne", "Nh", "Ni", "No", "Np",
  "O", "Og", "Os", "P", "Pa", "Pb", "Pd", "Pm", "Po", "Pr", "Pt", "Pu",
  "Ra", "Rb", "Re", "Rf", "Rg", "Rh", "Rn", "Ru", "S", "Sb", "Sc", "Se",
  "Sg", "Si", "Sm", "Sn", "Sr", "Ta", "Tb", "Tc", "Te", "Th", "Ti", "Tl",
  "Tm", "Ts", "U", "V", "W", "Xe", "Y", "Yb", "Zn", "Zr"
]
# fmt: on

elements = [el.lower().encode() for el in ELEMENTS]


def search(rng, shift, char):
    upper = rng[1]
    lower = rng[0]
    while lower < upper:
        mid = int((lower + upper) / 2)
        if elements[mid][shift] < char:
            lower = mid + 1
        else:
            upper = mid

    if lower == rng[1] or elements[lower][shift] != char:
        rng[1] = 0
        return

    upper = rng[1]
    rng[0] = lower
    while lower < upper:
        mid = int((lower + upper) / 2)
        if char < elements[mid][shift]:
            upper = mid
        else:
            lower = mid + 1

    rng[1] = upper


def split(tail):
    result = []

    rng = [0, len(ELEMENTS)]
    shift = 0

    while shift < len(tail):
        search(rng, shift, tail[shift])
        if rng[0] >= rng[1]:
            break

        shift += 1
        if len(elements[rng[0]]) == shift:
            result.append((ELEMENTS[rng[0]], tail[shift:]))
            rng[0] += 1

    return result or [("?", tail[1:])]


def explode(tail):
    return [(x[0], explode(x[1]) if x[1] else None) for x in split(tail)]


def analyze(word):
    return explode(word.lower().encode())


def print_plain(tree, formula):
    for x in tree:
        formula.append(x[0])
        if x[1]:
            print_plain(x[1], formula)
        else:
            print(" " + " ".join(formula))
        formula.pop()


for w in sys.argv[1:]:
    print(w + ":")
    if w:
        print_plain(analyze(w), [])

File: mendeleev.rs
use std::{collections::LinkedList, env, ops::Range, str};

#[rustfmt::skip]
static ELEMENTS : [&[u8]; 119] = [ b"?",
  b"Ac", b"Ag", b"Al", b"Am", b"Ar", b"As", b"At", b"Au", b"B",  b"Ba", b"Be", b"Bh",
  b"Bi", b"Bk", b"Br", b"C",  b"Ca", b"Cd", b"Ce", b"Cf", b"Cl", b"Cm", b"Cn", b"Co",
  b"Cr", b"Cs", b"Cu", b"Db", b"Ds", b"Dy", b"Er", b"Es", b"Eu", b"F",  b"Fe", b"Fl",
  b"Fm", b"Fr", b"Ga", b"Gd", b"Ge", b"H",  b"He", b"Hf", b"Hg", b"Ho", b"Hs", b"I",
  b"In", b"Ir", b"K",  b"Kr", b"La", b"Li", b"Lr", b"Lu", b"Lv", b"Mc", b"Md", b"Mg",
  b"Mn", b"Mo", b"Mt", b"N",  b"Na", b"Nb", b"Nd", b"Ne", b"Nh", b"Ni", b"No", b"Np",
  b"O",  b"Og", b"Os", b"P",  b"Pa", b"Pb", b"Pd", b"Pm", b"Po", b"Pr", b"Pt", b"Pu",
  b"Ra", b"Rb", b"Re", b"Rf", b"Rg", b"Rh", b"Rn", b"Ru", b"S",  b"Sb", b"Sc", b"Se",
  b"Sg", b"Si", b"Sm", b"Sn", b"Sr", b"Ta", b"Tb", b"Tc", b"Te", b"Th", b"Ti", b"Tl",
  b"Tm", b"Ts", b"U",  b"V",  b"W",  b"Xe", b"Y",  b"Yb", b"Zn", b"Zr"
];

struct Split<'a> {
    eid: usize,
    tail: &'a [u8],
}

struct Element {
    eid: usize,
    next: Option<LinkedList<Element>>,
}

fn search(range: &mut Range<usize>, shift: usize, mut c: u8) {
    let mut u = range.end;
    let mut l = range.start;
    c |= b' ';

    while l < u {
        let m = (l + u) / 2;
        if (ELEMENTS[m][shift] | b' ') < c {
            l = m + 1;
        } else {
            u = m;
        }
    }

    if !range.contains(&l) || ((ELEMENTS[l][shift] | b' ') != c) {
        range.end = 0;
        return;
    }

    u = range.end;
    range.start = l;
    while l < u {
        let m = (l + u) / 2;
        if c < (ELEMENTS[m][shift] | b' ') {
            u = m;
        } else {
            l = m + 1;
        }
    }

    range.end = u;
}

fn split(tail: &[u8]) -> LinkedList<Split> {
    let mut x = LinkedList::new();
    let mut shift = 0;
    let mut range = 1..ELEMENTS.len() - 1;

    while shift < tail.len() {
        search(&mut range, shift, tail[shift]);
        if range.is_empty() {
            break;
        }
        shift += 1;
        if shift == ELEMENTS[range.start].len() {
            x.push_back(Split {
                eid: range.start,
                tail: &tail[shift..],
            });
            range.start += 1;
        }
    }

    if x.is_empty() {
        x.push_back(Split {
            eid: 0,
            tail: &tail[1..],
        });
    }

    x
}

fn explode(tail: &[u8]) -> LinkedList<Element> {
    split(tail)
        .into_iter()
        .map(|x| Element {
            eid: x.eid,
            next: if x.tail.is_empty() {
                None
            } else {
                Some(explode(x.tail))
            },
        })
        .collect()
}

fn print_plain(tree: &LinkedList<Element>, formula: &mut Vec<usize>) {
    for x in tree {
        formula.push(x.eid);
        if let Some(next) = &x.next {
            print_plain(next, formula);
        } else {
            for i in formula.iter() {
                print!(" {}", unsafe { str::from_utf8_unchecked(ELEMENTS[*i]) });
            }
            println!();
        }
        formula.pop();
    }
}

fn main() {
    for word in env::args().skip(1) {
        println!("{word}:");
        if !word.is_empty() {
            let tail = word.as_bytes();
            print_plain(&explode(tail), &mut Vec::with_capacity(tail.len()));
        }
    }
}
